sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -38,95 +260,208 @@ static kmac_error_t wait_status_bit(uint32_t bit_position, uint32_t bit_value) {
 }
 
 OT_WARN_UNUSED_RESULT
-kmac_error_t kmac_init(kmac_operation_t operation) {
+static kmac_error_t write_left_encode(const size_t len, uint8_t *encoded_len,
+                                      size_t *write_ctr) {
+  if (encoded_len == NULL || write_ctr == NULL || len > kKmacPrefixMaxSize) {
+    return kKmacArgsError;
+  }
+
+  size_t bit_len = len * 8;
+  if (bit_len < 256) {
+    // Need 1 byte to encode `len`
+    *encoded_len++ = 0x01;
+    *encoded_len++ = bit_len;
+    *write_ctr += 2;
+  } else {
+    // Need 2 bytes to encode `len`
+    *encoded_len++ = 0x02;
+    *encoded_len++ = bit_len / 256;
+    *encoded_len++ = bit_len % 256;
+    *write_ctr += 3;
+  }
+  // `bit_len` is upper bounded by kKmacPrefixSize * 8 = 288, so we don't need
+  // to implement larger encoding cases
+
+  return kKmacOk;
+}
+
+/*
+ * Set prefix registers.
+ */
+OT_WARN_UNUSED_RESULT
+static kmac_error_t kmac_set_prefix_regs(const uint8_t *func_name,
+                                         const size_t func_name_len,
+                                         const uint8_t *cust_str,
+                                         const size_t cust_str_len) {
+  // static_assert(kKmacPrefixMaxSize < 8192, ");
+
+  uint32_t prefix_buffer[kKmacPrefixRegCount] = {0x0};
+  uint8_t *prefix_buf_ptr = (uint8_t *)prefix_buffer;
```

Comment:
Technically, this violates strict aliasing rules and is undefined behvior. The nicest way to get around this is just to use `memcpy`, which is written specifically with aliasing constraints in mind. (Or, in this case, you could probably just use `uint8_t` everywhere at the cost of a few extra writes to the prefix regs.) This is a bit of a recurring issue in crypto code for OpenTitan, since most hardware blocks expect 32-bit writes and a lot of crypto specs use byte buffers. I'm hoping to write some generic utilities that make it easier sometime soon.

Reply:
In the meantime, maybe crypto code should be compiled with `-fno-strict-aliasing`.

Reply:
That's a good thought. Just to make sure I understand my compiler flags correctly: is it accurate that `-fno-strict-aliasing` would make the compiler skip passes that assume we're following strict aliasing rules?

Reply:
Oh, I thought since we want to move away from violating those aliasing rules, we actually want to keep this warning flags on until we address them all (so that it keeps bugging us to fix it and we don't miss them).

Reply:
@jadephilipoom Yep, my understanding is that the compiler will skip strict aliasing optimization passes when it sees `-fno-strict-aliasing`. Here's a GCC doc that describes the positive variant of the flag (Clang's docs are a little lacking): https://gcc.gnu.org/onlinedocs/gcc-4.9.1/gcc/Optimize-Options.html

@ballifatih I've had trouble getting Clang and GCC to warn about strict aliasing violations — have you been getting any warnings? They typically just silently optimize in my experience. I think passing this flag would just reduce the danger of subtly broken code, not make it harder to find code that breaks the strict aliasing assumption.

Reply:
@dmcardle I didn't get any warnings so far. At least I don't see anything when I run `bazel build //sw/device/lib/crypto/drivers:kmac_test_vectors` (or when I pass `--verbose_failures` as well).

I see your point now, I actually misinterpreted this option as a warning flag, instead of an optimization flag. I think it is a good idea to pass `-fno-strict-aliasing` flag until we address memory aliasing. What is the correct way of passing this parameter in the build file, is it something like this?
```
copts = [
        "-fno-strict-aliasing",
    ]
```

Reply:
Yep, that's what I was imagining! If you do that, I think it would make sense to define a COPTS variable in the BUILD file and pass that into each target, rather than repeating the option each time.

There might be a clever way to augment the current toolchain's copts for targets in a particular package, but my Bazel skills are not strong enough. Some searching turned up this issue, but it doesn't offer any immediate solutions: https://github.com/bazelbuild/bazel/issues/11280

Actually, I wonder if we should consider turning off strict aliasing optimizations for all of OpenTitan. If it doesn't increase code size too much, I would personally consider it a correctness improvement.

Reply:
I'll file an issue about this on Monday. Unless you beat me to it, @ballifatih :)

Reply:
I opened the issue #16556, and feel free to take over the discussion @dmcardle. I don't think I fully understand the context here regarding memory aliasing violations. Thanks for the suggestion :)